sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala

```diff
@@ -515,28 +515,66 @@ case class AlterTableRenamePartitionCommand(
  */
 case class AlterTableDropPartitionCommand(
     tableName: TableIdentifier,
-    specs: Seq[TablePartitionSpec],
+    partitions: Seq[(TablePartitionSpec, Seq[Expression])],
     ifExists: Boolean,
     purge: Boolean,
     retainData: Boolean)
-  extends RunnableCommand {
+  extends RunnableCommand with PredicateHelper {
 
   override def run(sparkSession: SparkSession): Seq[Row] = {
     val catalog = sparkSession.sessionState.catalog
     val table = catalog.getTableMetadata(tableName)
+    val resolver = sparkSession.sessionState.conf.resolver
     DDLUtils.verifyAlterTableType(catalog, table, isView = false)
     DDLUtils.verifyPartitionProviderIsHive(sparkSession, table, "ALTER TABLE DROP PARTITION")
 
-    val normalizedSpecs = specs.map { spec =>
-      PartitioningUtils.normalizePartitionSpec(
-        spec,
-        table.partitionColumnNames,
-        table.identifier.quotedString,
-        sparkSession.sessionState.conf.resolver)
+    val toDrop = partitions.flatMap { partition =>
+      val normalizedSpecs = PartitioningUtils.normalizePartitionSpec(
+            partition._1,
+            table.partitionColumnNames,
+            table.identifier.quotedString,
+            sparkSession.sessionState.conf.resolver)
+
+      val partitionSet = {
+        if (partition._2.nonEmpty) {
+          val parts = partition._2.map { expr =>
+            val (attrName, value) = expr match {
+              case BinaryComparison(UnresolvedAttribute(name :: Nil), constant: Literal) =>
+                (name, constant.value)
+            }
+            if (!table.partitionColumnNames.exists(resolver(_, attrName))) {
+              throw new AnalysisException(s"${attrName} is not a valid partition column " +
+                s"in table ${table.identifier.quotedString}.")
+            }
+            val dataType = table.partitionSchema.apply(attrName).dataType
+            expr.withNewChildren(Seq(AttributeReference(attrName, dataType)(),
+              Cast(Literal(value.toString), dataType)))
+          }.reduce(And)
+
+          val partitions = catalog.listPartitionsByFilter(
+            table.identifier, Seq(parts)).map(_.spec)
+          if (partitions.isEmpty && !ifExists) {
```

Comment:
why do we need this check here?

Reply:
there are two occasions if we get a empty seq here from expression filters in one `DROP PARTITION` sql.
1. there is at least one filter but there is no partitions for the filter.
2. there is no filters

if we don't add this check, this may be confusing later.
because in the first occasion, we should use the `intersect` with `normalizedPartitionSpec`
but in the second occasion, we shouldn't use `intersect` because that will return a empty result.

add this check and we can treat them with different ways
1. regardless of `normalizedPartitionSpec` and throw an exception directly
2. return `Seq.empty`

Reply:
sorry, I don't really get what you mean. If we have no filters we are returning and empty `Seq` (the check is at line 539). So here we are in the case 1, ie. there is a filter and it returns no partitions. If we avoid this if, my understanding is that we return `partitions` - which is empty - to `partitionSet`. Then `toDrop` also would be empty. The result is that we call `dropPartitions` with an empty Seq and it will throw the `AnalysisException` (instead of doing it here). So I think this is useless. Am I wrong?

PS all these operations are becoming quite complex as inline statements. I think that creating some methods for handling the different parts could improve readability. What do you think?

Reply:
let me explain these two occasions more clearly. two sqls for example(the useless_expression means there are no partitions for the expression):
`ALTER TABLE DROP PARTITION(partitionVal1, useless_expression)`
`ALTER TABLE DROP PARTITION(partitionVal1)`

the first sql should drop partition `partitionVal1` <b>intersect</b> `useless_expression`, and it's empty.
the second sql should drop partition `partitionVal1`

if we return `Seq.empty` to `partitonSet`  for both sqls, it will be impossible to tell between them later.

Reply:
yes, but in the first case `toDrop` would be empty, in the second case it would contain `partitionVal1`. So when it is passed later to `dropPartitions`, this method checks if it is empty or not.

Reply:
I'm a little confusing. if we return `Seq.empty` for both cases to `partitionSet`. then the code will both go to line 570. 
how can we return empty for the first case to `toDrop` and return `partitionVal1` for the second case at this line.

Reply:
oh I see now.... well this is getting very involved... can we split the cases in different methods? I think we can have the 4 cases like:
```
 if (partition._1.isEmpty && !partition._2.isEmpty){
  // extract from partition._2
} else if (!partition._1.isEmpty && partition._2.isEmpty) {
  // extract from partition._2
} else if (!partition._1.isEmpty && !partition._2.isEmpty) {
  // intersect
} else {
  // return empty seq
}
```

Maybe with some comments to explain when each of these cases can happen. Thanks.

Reply:
Sure. I will make these codes more readable.